projects/24_push_constants/PushConstants.cpp

```diff
@@ -0,0 +1,369 @@
+
+            -1.0f, 1.0f,-1.0f,   1.0f, 0.0f,  // +Y side
+            -1.0f, 1.0f, 1.0f,   0.0f, 0.0f,
+             1.0f, 1.0f, 1.0f,   0.0f, 1.0f,
+            -1.0f, 1.0f,-1.0f,   1.0f, 0.0f,
+             1.0f, 1.0f, 1.0f,   0.0f, 1.0f,
+             1.0f, 1.0f,-1.0f,   1.0f, 1.0f,
+        };
+        // clang-format on
+        uint32_t dataSize = ppx::SizeInBytesU32(vertexData);
+
+        grfx::BufferCreateInfo bufferCreateInfo       = {};
+        bufferCreateInfo.size                         = dataSize;
+        bufferCreateInfo.usageFlags.bits.vertexBuffer = true;
+        bufferCreateInfo.memoryUsage                  = grfx::MEMORY_USAGE_CPU_TO_GPU;
+
+        PPX_CHECKED_CALL(GetDevice()->CreateBuffer(&bufferCreateInfo, &mVertexBuffer));
+
+        void* pAddr = nullptr;
+        PPX_CHECKED_CALL(mVertexBuffer->MapMemory(0, &pAddr));
+        memcpy(pAddr, vertexData.data(), dataSize);
+        mVertexBuffer->UnmapMemory();
+    }
+}
+
+void PushConstantsApp::Render()
+{
+    PerFrame& frame = mPerFrame[0];
+
+    grfx::SwapchainPtr swapchain = GetSwapchain();
+
+    uint32_t imageIndex = UINT32_MAX;
+    PPX_CHECKED_CALL(swapchain->AcquireNextImage(UINT64_MAX, frame.imageAcquiredSemaphore, frame.imageAcquiredFence, &imageIndex));
+
+    // Wait for and reset image acquired fence
+    PPX_CHECKED_CALL(frame.imageAcquiredFence->WaitAndReset());
+
+    // Wait for and reset render complete fence
+    PPX_CHECKED_CALL(frame.renderCompleteFence->WaitAndReset());
+
+    // Build command buffer
+    PPX_CHECKED_CALL(frame.cmd->Begin());
+    {
+        grfx::RenderPassPtr renderPass = swapchain->GetRenderPass(imageIndex);
+        PPX_ASSERT_MSG(!renderPass.IsNull(), "render pass object is null");
+
+        grfx::RenderPassBeginInfo beginInfo = {};
+        beginInfo.pRenderPass               = renderPass;
+        beginInfo.renderArea                = renderPass->GetRenderArea();
+        beginInfo.RTVClearCount             = 1;
+        beginInfo.RTVClearValues[0]         = {{0, 0, 0, 0}};
+        beginInfo.DSVClearValue             = {1.0f, 0xFF};
+
+        frame.cmd->TransitionImageLayout(renderPass->GetRenderTargetImage(0), PPX_ALL_SUBRESOURCES, grfx::RESOURCE_STATE_PRESENT, grfx::RESOURCE_STATE_RENDER_TARGET);
+        frame.cmd->BeginRenderPass(&beginInfo);
+        {
+            frame.cmd->SetScissors(GetScissor());
+            frame.cmd->SetViewports(GetViewport());
+            frame.cmd->BindGraphicsPipeline(mPipeline);
+            frame.cmd->BindVertexBuffers(1, &mVertexBuffer, &mVertexBinding.GetStride());
+
+            // Bind descriptor set
+            frame.cmd->BindGraphicsDescriptorSets(mPipelineInterface, 1, &mDescriptorSet);
+
+            // Get elapsed seconds
+            float t = GetElapsedSeconds();
+
+            // Draw center cube
+            {
+                // Calculate MVP
+                float4x4 P   = glm::perspective(glm::radians(60.0f), GetWindowAspect(), 0.001f, 10000.0f);
```

Comment:
move P and V as const in the scope above?

Reply:
Done.